README.md
# logicheck
Logicheck is an interactive tool to test logical structures and expressions. It can be used to determine the validity of deductive arguments. It can also display the truth tables for such arguments, or just for a set of logical expressions. It is suitable for users familiar with either propositional logic or Boolean logic.

ChangeLog
Version history:

2.1 (2016-07-24)

Extended functionality of "back" button to include conclusions.

Removed ability to change the most recently added premise to a conclusion,
due to the above change, and an associated bug.

Added the ability to generate and view truth tables for any set of logical
expressions.

2.0 (2016-07)

Migrated from TkInter to PyQt GUI package.

1.1 (2015)

Added "back" button that allows the most recent added premise to be removed
and returned to the entry line.

1.0 (2015)

Initial release with the core functionality that remains largely unchanged up
to version 2.1. Included the ability to add a series of premises (logical
expressions) and a conclusion to form a deductive argument. The program would
then determine the logical validity of the argument; if invalid, a list of
counterexamples that provided proof of invalidity was shown. At any time,
the user could reset to return the program to its initial state.

truth_table.py
#!/usr/bin/env python

"""
ArgTab.py

Copyright 2016 Ben Cottier

This file is part of Logicheck.

Logicheck is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Logicheck is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Logicheck.  If not, see <http://www.gnu.org/licenses/>.
"""

from PyQt5.QtWidgets import *
from PyQt5.QtGui import QIcon


class TruthTable(object):
    """Handles the display of truth tables and the associated data.
    """

    def __init__(self, symbols, premises, symbol_truth, premise_truth,
                 conc=True):
        """
        Constructor

        __init__(list<str>, list<str>, list<str>, list<list<int>>, bool)
        """

        self.symbols = symbols
        self.premises = premises
        self.symbol_truth = symbol_truth
        self.premise_truth = premise_truth
        self.conc = conc

        self.generate_table_data()

    def generate_table_data(self):
        """Processes the data for the table and sorts it into a list mirroring
        the table structure.
        """

        # Convert elements of premise_truth from int to str in a new list.
        self.conversion = []
        for i in range(len(self.premise_truth)):
            self.sub_list = []
            for j in range(len(self.premise_truth[i])):
                self.sub_list.append(str(self.premise_truth[i][j]))
            self.conversion.append(self.sub_list)

        l = len(self.premises)
        premises2 = []

        # Format logical expressions for column headings in a new list.
        for i in range(l):
            if i == l - 1 and self.conc:
                # Prepend a "therefore" symbol to the conclusion.
                premises2.append("".join([u'\u2234', "    ",
                                          self.premises[-1]]))
            else:
                # Prepend a number to the expression.
                premises2.append("".join([str(i+1), ".    ",
                                          self.premises[i]]))

        # First row of table data - symbols then premises as column headings.
        self.table_data = [self.symbols]
        self.table_data[0].extend(premises2)

        # Create the remaining rows of truth values, i.e. 1's and 0's.
        for i in range(len(self.symbol_truth)):
            self.table_row = []
            for j in range(len(self.symbol_truth[i])):
                self.table_row.append(self.symbol_truth[i][j])
            self.table_row.extend(self.conversion[i])
            self.table_data.append(self.table_row)

        # E.g.
        # symbol_truth = [('0', '0', '1'), ...]
        # premise_truth = [[1, 0, 1], ...]
        # table = [ ['A', 'B', 'C', 'A->B', 'C+A', 'B'],
        # ['0', '0', '1', '1', '0', '1'], ...]

    def get_table_data(self):
        """Returns the table data, as produced by generate_table_data().
        """
        return self.table_data


class TruthTableGraphic(QTableWidget):
    """The widget used as the table to be displayed.
    """

    def __init__(self, table_data):
        """
        Constructor

        __init__(list<list<str>>)
        """

        # Inherit from QTableWidget.
        super().__init__()
        self.table_data = table_data
        # Make the table read-only.
        self.setEditTriggers(QAbstractItemView.NoEditTriggers)
        self.setRowCount(len(self.table_data)-1)
        self.setColumnCount(len(self.table_data[0]))
        self.setHorizontalHeaderLabels(self.table_data[0])
        # Table styling - not yet working.
        # stylesheet = "QHeaderView::section{" \
        #              "background-color:Whitesmoke;" \
        #              "}"
        # self.setStyleSheet(stylesheet)

        # Fill out the rest of the table with truth values.
        for i in range(1, len(self.table_data)):
            for j in range(len(self.table_data[i])):
                item = QTableWidgetItem(self.table_data[i][j])
                # Centre.
                item.setTextAlignment(132)
                # Add table item to table.
                self.setItem(i-1, j, item)

        self.resizeColumnsToContents()
        self.resizeRowsToContents()
        # self.resize(self.sizeHint())
        # self.setFixedSize(self.horizontalHeader().length() + 40,
        #                   self.verticalHeader().length() + 50)


class TruthTableWindow(QMainWindow):
    """The window that displays the truth table.
    """

    def __init__(self, table_data):
        """
        Constructor

        __init__(list<list<str>>)
        """

        # Inherit from QMainWindow.
        super().__init__()
        main_layout = QGridLayout()
        truth_table_graphic = TruthTableGraphic(table_data)
        # window_width = truth_table_graphic.width()
        # window_height = truth_table_graphic.height()
        main_layout.addWidget(truth_table_graphic, 0, 0)
        main_widget = QWidget()
        main_widget.setLayout(main_layout)

        self.setCentralWidget(main_widget)
        # Place this window slightly offset from the starting position of the
        # main window.
        self.setGeometry(80, 110, 425, 540)
        self.setWindowTitle("Logicheck - Truth Table")
        self.setWindowIcon(QIcon("LogicheckIcon3.png"))


logicheck.py
#!/usr/bin/env python

"""
logicheck.py

Copyright 2016 Ben Cottier

This file is part of Logicheck.

Logicheck is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Logicheck is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Logicheck.  If not, see <http://www.gnu.org/licenses/>.
"""

import sys

from PyQt5.QtWidgets import *
from PyQt5.QtGui import QIcon, QFont

from argument import ArgCheck


class Logicheck(QMainWindow):
    """The main window for the application.
    """

    def __init__(self):
        """
        Constructor
        """

        # Inherit from QMainWindow.
        super().__init__()

        # Top-level config:
        self.setCentralWidget(ArgCheck(self))
        # Set relative window position (first two args) and size.
        self.setGeometry(50, 80, 425, 540)
        self.setWindowTitle("Logicheck")
        self.setWindowIcon(QIcon("images/logicheck_icon_3.png"))

        # Create menu.
        # Create a help action which will go in a drop-down menu.
        # The "&" in "&Help" specifies the Alt- shortcut key.
        helpAction = QAction(QIcon("images/logicheck_help_icon.png"),
                             "&Help", self)
        helpAction.setShortcut("Ctrl+H")
        # Connect the selection of the action to displaying the manual
        helpAction.triggered.connect(self.show_info)
        # Create the status bar and the menu bar
        self.statusBar()
        menubar = self.menuBar()
        helpMenu = menubar.addMenu("&Help")
        # Link the menu item to the action.
        helpMenu.addAction(helpAction)

    def show_info(self):
        # Create window displaying the manual.
        info_file = open("documents/manual.html")
        # Get the contents (text).
        info = info_file.read()
        self.info_window = ManualWindow(info)
        font1 = QFont()
        font1.setPointSize(11)
        self.info_window.setFont(font1)
        self.info_window.show()


class ManualWindow(QMainWindow):
    """The window for displaying the operation manual.
    """

    def __init__(self, info):
        """
        Constructor
        """

        super().__init__()

        # Create object to display text.
        info_text = QTextEdit()
        info_text.setReadOnly(True)
        info_text.setText(info)
        # Create layout to store text object.
        info_layout = QGridLayout()
        info_layout.addWidget(info_text, 0, 0)
        info_widget = QWidget()
        info_widget.setLayout(info_layout)

        self.setCentralWidget(info_widget)
        self.setGeometry(80, 110, 425, 540)
        self.setWindowTitle("Logicheck - Manual")
        self.setWindowIcon(QIcon("images/logicheck_icon_3.png"))

# Create window.
# Execute the program only if the file was run directly, not imported.
if __name__ == "__main__":
    # Create application object.
    # "sys.argv" is a list of args from the command line.
    # This allows controlled startup from the shell.
    app = QApplication(sys.argv)
    # Initialise main window and display it.
    window = Logicheck()
    window.show()
    # Enter the mainloop, which exits if exit() is called or the main widget
    # is destroyed. This gives a clean exit, with exit code.
    sys.exit(app.exec_())
